From 91fb354f354c48610de92388e94c408f606e05d1 Mon Sep 17 00:00:00 2001
From: Tianmi Chen <tianmi.chen@intel.com>
Date: Tue, 30 Apr 2013 14:20:33 +0800
Subject: libstagefright: Set sync flag in case the frame is a sync frame

BZ: 50163

Set sync flag in case the frame is a sync frame,
indicating current frame can be decoded independently
when seeking.

Category: aosp improvement
Domain: Video.Media-local playback
Origin: internal
Upstream-Candidate: yes

Change-Id: I95e54ad53942553ae378d4b8de51abd8c8fd1bf1
Orig-Change-Id: I495f5948c7527f7c8e1f44712f19e3ff9ffbb74e
Signed-off-by: Tianmi Chen <tianmi.chen@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
Signed-off-by: Yuanjun Huang <yuanjun.huang@intel.com>
---
 include/media/stagefright/MetaData.h       |    1 +
 media/libstagefright/MPEG4Extractor.cpp    |    8 ++++++++
 media/libstagefright/OMXCodec.cpp          |    8 +++++++-
 media/libstagefright/SampleTable.cpp       |    4 ++++
 media/libstagefright/include/SampleTable.h |    2 ++
 5 files changed, 22 insertions(+), 1 deletions(-)

diff --git a/include/media/stagefright/MetaData.h b/include/media/stagefright/MetaData.h
index 92f2370..b82a4ca 100644
--- a/include/media/stagefright/MetaData.h
+++ b/include/media/stagefright/MetaData.h
@@ -58,6 +58,7 @@ enum {
     kKeyVorbisBooks       = 'vboo',  // raw data
     kKeyWantsNALFragments = 'NALf',
     kKeyIsSyncFrame       = 'sync',  // int32_t (bool)
+    kKeyIsFakeSync        = 'fake',  // int32_t (bool)
     kKeyIsCodecConfig     = 'conf',  // int32_t (bool)
     kKeyTime              = 'time',  // int64_t (usecs)
     kKeyDecodingTime      = 'decT',  // int64_t (decoding timestamp in usecs)
diff --git a/media/libstagefright/MPEG4Extractor.cpp b/media/libstagefright/MPEG4Extractor.cpp
index 2708f35..17a8d24 100644
--- a/media/libstagefright/MPEG4Extractor.cpp
+++ b/media/libstagefright/MPEG4Extractor.cpp
@@ -3185,6 +3185,10 @@ status_t MPEG4Source::read(
                 mBuffer->meta_data()->setInt32(kKeyIsSyncFrame, 1);
             }
 
+            if (!(mSampleTable->isSyncTableValid())) { /* no stss table */
+                mBuffer->meta_data()->setInt32(kKeyIsFakeSync, 1);
+            }
+
             ++mCurrentSampleIndex;
         }
 
@@ -3307,6 +3311,10 @@ status_t MPEG4Source::read(
             mBuffer->meta_data()->setInt32(kKeyIsSyncFrame, 1);
         }
 
+        if (!(mSampleTable->isSyncTableValid())) { /* no stss table */
+            mBuffer->meta_data()->setInt32(kKeyIsFakeSync, 1);
+        }
+
         ++mCurrentSampleIndex;
 
         *out = mBuffer;
diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 29945ce..dd9a002 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -3066,7 +3066,7 @@ bool OMXCodec::drainInputBuffer(BufferInfo *info) {
 
     size_t offset = 0;
     int32_t n = 0;
-
+    int32_t isSync = 0, isFakeSync = 0;
 
     for (;;) {
         MediaBuffer *srcBuffer;
@@ -3171,6 +3171,8 @@ bool OMXCodec::drainInputBuffer(BufferInfo *info) {
         }
 
         offset += srcBuffer->range_length();
+        srcBuffer->meta_data()->findInt32(kKeyIsSyncFrame, &isSync);
+        srcBuffer->meta_data()->findInt32(kKeyIsFakeSync, &isFakeSync);
 
         if (!strcasecmp(MEDIA_MIMETYPE_AUDIO_VORBIS, mMIME)) {
             CHECK(!(mQuirks & kSupportsMultipleFramesPerInputBuffer));
@@ -3220,6 +3222,10 @@ bool OMXCodec::drainInputBuffer(BufferInfo *info) {
         mNoMoreOutputData = false;
     }
 
+    if (isSync == 1 && !isFakeSync) {
+        flags |= OMX_BUFFERFLAG_SYNCFRAME;
+    }
+
     if (info == NULL) {
         CHECK(mFlags & kUseSecureInputBuffers);
         CHECK(signalEOS);
diff --git a/media/libstagefright/SampleTable.cpp b/media/libstagefright/SampleTable.cpp
index 4a55384..5f1bcd7 100644
--- a/media/libstagefright/SampleTable.cpp
+++ b/media/libstagefright/SampleTable.cpp
@@ -160,6 +160,10 @@ bool SampleTable::isValid() const {
         && mTimeToSample != NULL;
 }
 
+bool SampleTable::isSyncTableValid() const {
+    return mSyncSampleOffset >= 0;
+}
+
 status_t SampleTable::setChunkOffsetParams(
         uint32_t type, off64_t data_offset, size_t data_size) {
     if (mChunkOffsetOffset >= 0) {
diff --git a/media/libstagefright/include/SampleTable.h b/media/libstagefright/include/SampleTable.h
index 67f7008..3d1eab7 100644
--- a/media/libstagefright/include/SampleTable.h
+++ b/media/libstagefright/include/SampleTable.h
@@ -82,6 +82,8 @@ public:
 
     status_t findThumbnailSample(uint32_t *sample_index);
 
+    bool isSyncTableValid() const;
+
 protected:
     ~SampleTable();
 
-- 
1.7.4.1

